README.md
# Snake-Game-AI-Agent
A user interface with features like:
 - A simple snake game made in Python using the PyGame library
 - An AI agent that uses a genetic algorithm and fundamental concepts like crossover and mutation to learn to play snake.
 - The ability to play the agent in a score based snake game or an arena where your both playing at the same time.
 - Graphing displaying the training process.
 - A seamless transition from the AI menu to the main menu, gamemodes back to AI menu, 
 - Ability to go from gamemode back to menu without having to relaunch and retrain the agent.

## Installation

Use the package manager [pip](https://pip.pypa.io/en/stable/) to install required packages.

```bash
pip install pygame
pip install matplotlib
pip install numpy
pip install pygame
pip install scipy
```

## Usage

```python
python snake-game.py
```

## Contributing

Pull requests are welcome. For major changes, please open an issue first
to discuss what you would like to change.


CHANGELOG.md
# Snake-Game-AI-Agent

## 15 May 2023:
 - Finished the project on the 12th of May and have taken a few days off. I dove into this head first and ended up spending 6+ hours a day on it but enjoyed every second.
 - I changed the small, medium and large buttons to reflect steps instead of generations. This is because it usually takes around 75 generations regardsless of steps to train the generation so doing more / fewer was pointless. This makes it train faster with being a couple seconds for small (1500 steps/generation) and 5 - 6 seconds for large (10,000 steps/generation).
 - Added in deathmatch which allows you to vs the Agent in the same arena, both fighting for the same fruit. Win conditions are first to make the other run into their body or if one hits the wall they lose.
 - The current run time isn't too bad with the smaller steps being around 3 minutes and the larger being around 7.
 - A big problem is how intense it is on my system. It uses all avaliable CPU sometimes maxing out and while talking to a friend on discord he said my voice was cutting out while it was running. I also had Subnautica open while trying to train a generation and it crashed at around generation 30.
 - The graphics and colour scheme of the project isn't the best, but I don't see myself as the most creative person and definetly had more fun building the genetic algorithm then colouring / positioning buttons.
 - The actual program is working and I'm happy with it but I might do some research on processes and threads and see if I can get it more responsive in the training process.

## 9th May 2023:

### snake-game.py
 - I've now integrated the genetic algorithm and player controls into one file which is now the main file.
 - Sadly, I can't call an instance of AIAgentmultiprocessing.py because I can't share variables and return from one file to another since it's a whole different process and it becomes messy quick.
 - Have implemented a menu system where you can always go back and forth between the two options in the inital menu. This is so it's a smooth program to run.
 - Once the user has pressed "Start (with AI agent)" The code then creates another process for the training which is very computationally heavy. On the main process, it displays to the user information about the training process and once finished it allows the user to select from a range of options and gamemodes.
 - Have also implemented graphing while the population is training. There are two graphs, one for the current generation's fitness and another for max fitness in each generation. These are displayed to the user while the population trains. Also shows the user the generation number but is a bit laggy because of the sheer amount of computations happening in the background.
 - Training time is still 5 - 6 seconds with 2500 steps per generation and now has a variety of options to do after training is complete like:
     - Retrain if the agent for any reason.
     - View the agent playing the snake game.
     - Go head to head in a normal game of snake competing for the highest score.
     - --------------------------------------------------------- Not implemented yet ---------------------------------------------------------------
         * Will have a deathmatch gamemode where you and the agent are both in the same game fighting to hit each other and get the highest score.

## 4th May 2023:

Have recreated the game with the user controlls working and is in repository Snake-Game (https://github.com/aidanHorne978/Snake-Game).
      
### AIAgentmultiprocessing.py
 - Run's the same AI agent as AIagent.py but uses multiprocessing to speed up the generations.
 - Currently, I'm using a desktop with a AMD Ryzen 7 3700X 8-Core Processor 3.60 GHz CPU and I am running 50 Agents in a generation that each do 2,500 steps before returning. I've divided the population into 10 groups  of 5 which takes between 5 - 6 seconds to train a generation.
 - Comparing the 5 - 6 seconds with 2500 steps per generation to the 14 - 15 seconds it takes just using one process in AIagent.py, it is much more efficient and takes way less time (10 minutes instead of 25 minutes).

snakeTrain.py
#!/usr/bin/python
import pygame
import random
import numpy as np

# Screen resolution.
res = (900, 800)

# Grid size.
blockSize = 20

# Screen width, height and center
width = res[0]
height = res[1]

# Keep track of what direction.
distance = 0

# Snake class
class Snake:

    def __init__(self, x, y, body):
        self.x = x
        self.y = y
        self.body = body

    def pos(self):
        return self.x, self.y


class Fruit:

    def __init__(self, x, y):
        self.x = x
        self.y = y
    
    def pos(self):
        return self.x, self.y

    # Range is so it fits in the grid of the game.
    def generateFruit(self):
        newx = random.randrange(blockSize, res[0] - blockSize, blockSize)
        newy = random.randrange(blockSize, res[1] - blockSize - 100, blockSize)
        self.x = newx
        self.y = newy
        return self.x, self.y

def Grow(snake, direction):

    if direction == "left":
        segment = pygame.Rect(snake.body[-1].x + blockSize, snake.body[-1].y, blockSize, blockSize)
        snake.body.insert(len(snake.body), segment)
    if direction == "right":
        segment = pygame.Rect(snake.body[-1].x - blockSize, snake.body[-1].y, blockSize, blockSize)
        snake.body.insert(len(snake.body), segment)    
    if direction == "up":
        segment = pygame.Rect(snake.body[-1].x, snake.body[-1].y + blockSize, blockSize, blockSize)
        snake.body.insert(len(snake.body), segment)   
    if direction == "down":
        segment = pygame.Rect(snake.body[-1].x, snake.body[-1].y - blockSize, blockSize, blockSize)
        snake.body.insert(len(snake.body), segment)

def MoveSnake(direction, snake):

    # Moves the head of the snake.
    if direction == "left":
        snake.x -= blockSize
    elif direction == "right":
        snake.x += blockSize
    elif direction == "up":
        snake.y -= blockSize
    elif direction == "down":
        snake.y += blockSize

    # Code for the body.
    if len(snake.body) > 1:
        for j in reversed(range(1, len(snake.body))):

            if snake.body[j - 1] == snake.x:
                snake.body[j].x = snake.x
            else:
                snake.body[j].x = snake.body[j - 1].x

            if snake.body[j - 1] == snake.y:
                snake.body[j].y = snake.y
            else:
                snake.body[j].y = snake.body[j - 1].y

    # Update the head in the body list.
    snake.body[0].x = snake.x
    snake.body[0].y = snake.y

def SnakeGame(player, lastMove, fruit):

    ## Global variables.
    global fruits
    global score
    global distance

    if len(player.body) < 1:

        # Draws the grid for the game and the fruit.
        fruits = fruit
        fruits.generateFruit()
        distance = 0

        # Head of snake.
        player.body.append(pygame.Rect(player.x, player.y, blockSize, blockSize))

        # Tail of snake.
        player.body.append(pygame.Rect(player.x, player.y, blockSize, blockSize))

        # Score variable.
        score = 0

    MoveSnake(lastMove, player)
    distance += 1

    # When you get a fruit it will replace it with another randomly generated fruit.
    # Then it will add one to the score and display it.
    if player.x == fruits.pos()[0] and player.y == fruits.pos()[1]:

        # Adds a part to the body.
        Grow(player, lastMove)

        # Generate a new fruit.
        fruits.generateFruit()

        # Increase score by one.
        score += 1

    # If the snake hit's itself.
    if len(player.body) > 2:
        for parts in player.body[1:]:
            if player.x == parts.x and player.y == parts.y:
                return score, distance, False

    if player.x < 20 or player.x > 860:
        return score, distance, False

    elif player.y < 20 or player.y > 700:
        return score, distance, False
    else:
        return score, distance, True

old versions/AIagent.py
#!/usr/bin/python
import random
import numpy as np
import time
import os
os.environ['PYGAME_HIDE_SUPPORT_PROMPT'] = "hide"
import pygame

class Agent:

    def __init__(self, hScore, deaths, penalties, avg_steps):
        
        # Initilizing with the size of the player's chromosomes.
        self.chromosome = np.zeros(amountOfChromosomes)

        # Player's highest score in training.
        self.hScore = hScore

        # Amount of deaths in training.
        self.deaths = deaths

        # Penalties received during training.
        self.penalties = penalties

        # Average steps during training.
        self.avg_steps = avg_steps

    # Insert a chromosome at a certain index.
    def insert(self, value, index):
        self.chromosome[index] = value

    # Initilizes the chromosomes for a player.
    def createPopulation(self):
        self.chromosome = np.random.uniform(-1, 1, amountOfChromosomes)

class Snake:

    # Initilizes the snake with it's heads x, y value and body which is a list.
    def __init__(self, x, y, body):
        self.x = x
        self.y = y
        self.body = body

    # Draws the snake's body.
    def draw(self):
        return pygame.Rect(self.x, self.y, blockSize, blockSize)

class Fruit:

    # Initilizes the fruit object and which has an x and y coordinate.
    def __init__(self, x, y):
        self.x = x
        self.y = y

    # Generates a fruit with a range so it fits inside grid of the game.
    def generateFruit(self):
        newx = random.randrange(blockSize, res[0] - blockSize, blockSize)
        newy = random.randrange(blockSize, res[1] - blockSize - 100, blockSize)
        self.x = newx
        self.y = newy
        return self.x, self.y

    # Draw's the fruit on the grid.
    def drawFruit(self):
        rect = pygame.Rect(self.x, self.y, blockSize, blockSize)
        return pygame.draw.rect(screen, RED, rect)

def Grow(snake, direction):

    # Will create the new segment on the tile the snake's head just left.
    if direction == "left":
        segment = pygame.Rect(snake.body[-1].x + blockSize, snake.body[-1].y, blockSize, blockSize)
        snake.body.insert(len(snake.body), segment)
    if direction == "right":
        segment = pygame.Rect(snake.body[-1].x - blockSize, snake.body[-1].y, blockSize, blockSize)
        snake.body.insert(len(snake.body), segment)    
    if direction == "up":
        segment = pygame.Rect(snake.body[-1].x, snake.body[-1].y + blockSize, blockSize, blockSize)
        snake.body.insert(len(snake.body), segment)   
    if direction == "down":
        segment = pygame.Rect(snake.body[-1].x, snake.body[-1].y - blockSize, blockSize, blockSize)
        snake.body.insert(len(snake.body), segment)

def MoveSnake(direction, snake, draw):

    # Moves the head of the snake.
    if direction == "left":
        snake.x -= blockSize
    elif direction == "right":
        snake.x += blockSize
    elif direction == "up":
        snake.y -= blockSize
    elif direction == "down":
        snake.y += blockSize

    # If we are displaying to the user we draw it to the screen.
    if draw:
        pygame.draw.rect(screen, BLACK, snake.draw())

    # This moves the body in the way a snake does, by following the segment before it.
    if len(snake.body) > 1:

        for j in reversed(range(1, len(snake.body))):

            if snake.body[j - 1] == snake.x:
                snake.body[j].x = snake.x
            else:
                snake.body[j].x = snake.body[j - 1].x

            if snake.body[j - 1] == snake.y:
                snake.body[j].y = snake.y
            else:
                snake.body[j].y = snake.body[j - 1].y

            # If we are displaying to the user we draw it to the screen.
            if draw:
                pygame.draw.rect(screen, BLACK, snake.body[j])
                pygame.draw.rect(screen, pygame.Color(GRASS), snake.body[-1])
                pygame.draw.rect(screen, BLACK, snake.body[-1], 1)

    # Update the head in the body list.
    snake.body[0].x = snake.x
    snake.body[0].y = snake.y

def SnakeGame(agent, lastMove, fruit):

    scored = False

    # If this is the first run through for the current agent we initilize some variables.
    if len(agent.body) < 1:

        global score
        score = 0

        # Head of snake.
        agent.body.append(pygame.Rect(agent.x, agent.y, blockSize, blockSize))

        # Tail of snake.
        agent.body.append(pygame.Rect(agent.x, agent.y, blockSize, blockSize))

        # Score variable.
        score = 0

    # Move the snake in a direction picked by the neural network.
    MoveSnake(lastMove, agent, False)

    # When you get a fruit it will replace it with another randomly generated fruit.
    # Then it will add one to the score and display it.
    if agent.x == fruit.x and agent.y == fruit.y:

        # Adds a part to the body.
        Grow(agent, lastMove)

        # Increase score by one.
        score += 1
        scored = True

    # If the snake hit's itself.
    if len(agent.body) > 2:
        for parts in agent.body[1:]:
            if agent.x == parts.x and agent.y == parts.y:
                return score, scored, False

    # If the snake hit's the horizontal barrier we return with False which indicates a death.
    if agent.x < 20 or agent.x > 860:
        return score, scored, False

    # If the snake hit's the vertical barrier we return with False which indicates a death.
    if agent.y < 20 or agent.y > 700:
        return score, scored, False
    else:

        # If the snake didn't die then we return the score, whether they scored or not and True indicating it's still alive.
        return score, scored, True

def sigmoid(x):
    
    # Activation function.
    return 1 / (1 + np.exp(-x))

def make_prediction(input_vector, input_to_layer1, layer1_to_layer2, layer2_to_output, bias):

    # Neural network for the agent to make decisions.
    layer1 = np.dot(input_vector, input_to_layer1)
    layer2 = sigmoid(np.dot(layer1, layer1_to_layer2) + bias)
    output_layer = np.dot(layer2, layer2_to_output) + bias
    return output_layer

def makeMove(agent, snake, lastMove, fruit, debug):

    # --------------------------------- DEBUGGING ---------------------------------

    # print("Snake x: {}\tSnake y: {}\tFruit x: {}\tFruit y: {}".format(snake.x, snake.y, fruit.x, fruit.y))

    # --------------------------------- DEBUGGING ---------------------------------

    ## Input nodes ##

    # Orientation is Left, Right, Up, Down in the lists.

    # Read's a 1 if the food is in that direction, else 0. Will have 2 1's unless the snake is on the correct horizontal or vertical axis.
    posFood = [0, 0, 0, 0]
    if snake.x != fruit.x:
        if snake.x > fruit.x:
            posFood[0] = 1
        elif snake.x < fruit.x:
            posFood[1] = 1
    if snake.y != fruit.y:
        if snake.y > fruit.y:
            posFood[2] = 1
        elif snake.y < fruit.y:
            posFood[3] = 1

    # Read's 1 if there is a wall in that direction, else 0.
    wall = [0, 0, 0, 0]
    if snake.x == 20:
        wall[0] = 1
    if snake.x == 860:
        wall[1] = 1
    if snake.y == 20:
        wall[2] = 1
    if snake.y == 700:
        wall[3] = 1

    # Gives the relative position of the food like posFood. Index 0 will have -1 if food is left, else 0. Index 1 will have -1 if food is right, else 0.
    # Index 2 will have -1 if food is up, else 0, Index 3 will have -1 if food is down, else 0.
    # relativePos = [0, 0, 0, 0]
    # if fruit.x < snake.x:
    #     relativePos[0] = -1
    # elif fruit.x > snake.x:
    #     relativePos[1] = 1
        
    # if fruit.y < snake.y:
    #     relativePos[2] = -1
    # elif fruit.y > snake.y:
    #     relativePos[3] = 1

    # Current direction the snake is moving aka lastMove variable.
    currdirection = [0, 0, 0, 0]
    if lastMove == "left":
        currdirection[0] = 1
    if lastMove == "right":
        currdirection[1] = 1
    if lastMove == "up":
        currdirection[2] = 1
    if lastMove == "down":
        currdirection[3] = 1

    # Putting the input nodes into an array.
    input_vector = np.array(posFood[0])

    for i in range(4):
        if i > 0:
            input_vector = np.append(input_vector, posFood[i])
        input_vector = np.append(input_vector, wall[i])
        input_vector = np.append(input_vector, currdirection[i])
        # input_vector = np.append(input_vector, relativePos[i])


    # Neural network's structure is an input layer (12 -> 120) values, hidden layer (120 -> 120) values and an output layer (120 -> 4) values.
    input_to_layer1 = agent.chromosome[0:1440].reshape(12, 120)
    input_to_layer1[:, 0] = input_vector
    layer1_to_layer2 = agent.chromosome[1441:15841].reshape(120, 120)
    layer2_to_output = agent.chromosome[15841:16321].reshape(120, 4)

    # --------------------------------- DEBUGGING ---------------------------------
    # 14 input nodes.
    # input_to_layer1 = agent.chromosome[0:1680].reshape(14, 120)
    # input_to_layer1[:, 0] = input_vector
    # layer1_to_layer2 = agent.chromosome[1681:16081].reshape(120, 120)
    # layer2_to_layer3 = agent.chromosome[16082:30482].reshape(120, 120)
    # layer3_to_output = agent.chromosome[30483:30963].reshape(120, 4)

    # 16 input nodes
    # input_to_layer1 = agent.chromosome[0:1920].reshape(16, 120)
    # input_to_layer1[:, 0] = input_vector
    # layer1_to_layer2 = agent.chromosome[1921:16321].reshape(120, 120)
    # layer2_to_output = agent.chromosome[16321:16801].reshape(120, 4)

    if debug:
        return move, [wall, currdirection, posFood]

    # --------------------------------- DEBUGGING ---------------------------------

    # Return's a numpy array of 4 numbers with the highest being the best move.
    move = make_prediction(input_vector, input_to_layer1, layer1_to_layer2, layer2_to_output, 0)

    # Convert to list.
    move = move.tolist()
    return move

def runAgent(agent):

    # Initlizing variables needed for each individual agent.
    steps = 0
    firstFruit = 0
    first = True
    lastMove = "left"
    penalty = True

    fruit = Fruit(0, 0)
    aiSnake = Snake(random.randrange(20, 860, 20), random.randrange(20, 700, 20), [])
    fruit.generateFruit()

    # Creates a new snake in a random position with a new fruit aswell to promote dynamic behaviour.
    def respawn():
        newSnake = Snake(random.randrange(20, 860, 20), random.randrange(20, 700, 20), [])
        fruit = Fruit(0, 0)
        fruit.generateFruit()
        return newSnake

    # Train the agent for n number of steps.
    while True:
        
        # If the snake dies it then gets reset
        if aiSnake.x < 20 or aiSnake.x > 860:
            aiSnake = respawn()
            penalty = True
        elif aiSnake.y < 20 or aiSnake.y > 700:
            aiSnake = respawn()
            penalty = True

        # Runs the neural network to decide which direction to move.
        move = makeMove(agent, aiSnake, lastMove, fruit, False)

        # The highest value is the move the neural network thinks is the best.
        if move[0] == max(move):
            if lastMove != "right":
                lastMove = "left"
        elif move[1] == max(move):
            if lastMove != "left":
                lastMove = "right"
        elif move[2] == max(move):
            if lastMove != "down":
                lastMove = "up"
        elif move[3] == max(move):
            if lastMove != "up":
                lastMove = "down"

        # Run's the code for the snake game.
        evaluation = SnakeGame(aiSnake, lastMove, fruit)

        # To counter-act spinning by killing the snake if it spins and spawning in a new one.
        if steps % 200 == 0 and steps != 0 and penalty == True:
            if penalty == True:
                agent.penalties += 1
            agent.deaths += 1
            aiSnake = respawn()
            penalty = True

        # If the snake gets a fruit then we start counting steps.
        if evaluation[1]:
            if first:
                firstFruit = steps
                first = False
            fruit = Fruit(0, 0)
            fruit.generateFruit()
            agent.avg_steps.append(steps - firstFruit)
            firstFruit = steps
            penalty = False

        # Keeps track of deaths.
        if evaluation[2] == False:
            if evaluation[0] > agent.hScore:
                agent.hScore = evaluation[0]
            agent.deaths += 1
            first = True


        # Return the agent after training.
        if steps == 2500:
            return
            
        steps += 1

def fitness(population):

    scores = []
    # Review how the agent performed in that generation and return it's score.
    for agent in population:
        if agent.hScore > 0:
            score = agent.hScore * 5000  - agent.deaths * 150 - int(sum(agent.avg_steps) / len(agent.avg_steps) * 100) - agent.penalties * 1000
        else:
            score = agent.hScore * 5000  - agent.deaths * 150 - agent.penalties * 1000
        scores.append((score, agent))
    return scores

def crossover(population):

    # Split the 12 best from last generation into 2 lists so we don't crossover from the same parent.
    mother = population[0:5]
    father = population[6:11]
    newGeneration = []

    # Will crossover from the best 12 agents in the previous generation to create the next generation.
    while len(newGeneration) < 50:
        child = Agent(0, 0, 0, [])
        for j in range(amountOfChromosomes):

            coin = random.getrandbits(1)
            parent = random.randint(0, len(mother) - 1)

            if coin == 0:
                child.insert(mother[parent].chromosome[j], j)
            else:
                child.insert(father[parent].chromosome[j], j)

        newGeneration.append(child)

    return newGeneration

def mutation(population):

    prob = random.randint(0, 100)

    # Mutate the next generation at random to promote diversity and try avoid the local minima.
    if prob % 48 == 0:
        for j in range(random.randint(0, 24), random.randint(25, 50)):
            for i in range(len(population[j].chromosome)):
                population[j].chromosome[i] = population[j].chromosome[i] + random.uniform(-1, 1)
        print("mutated")
    
    return population

def trainGen(population, numGens):

    generation = 0

    # Train the n number of agents for x number of generations.
    while True:

        # Record how long each generation takes for debugging and efficiency purposes.
        start = time.time()

        # Priting out the current generation.
        print()
        print("Generation: {}".format(generation))
        
        newGeneration = []
        
        # Running the game for every agent in the generation.
        for i in range(50):

            # Multiprocess this.
            runAgent(population[i])
        
        # Finding the fitness for the generation and sorting by score to find best agents.
        currentFitness = fitness(population)
        currentFitness = sorted(currentFitness, key=lambda x: x[0], reverse=True)

        # This is just for displaying the current generation in one list.
        cFitness = []
        for x in currentFitness:
            cFitness.append(x[0])

        # Printing the generation's score and current maximum fitness score.
        print()
        print(cFitness)
        print()
        print("Current Max: {}".format(currentFitness[0][0]))
        print()

        # --------------------------------- DEBUGGING ---------------------------------

        # if generation > 90:
        #     for i in range(3):
        #         print("Score: {}".format(currentFitness[i][1].hScore))
        #         print("Deaths: {}".format(currentFitness[i][1].deaths))
        #         print("Penaltys: {}".format(currentFitness[i][1].penalties))
        #         print("Average steps: {}".format(currentFitness[i][1].avg_steps))
        #         print()
        #         displayGame(Snake(random.randrange(20, 860, 20), random.randrange(20, 700, 20), []), "left", population[i], Fruit(0, 0))
        #     pygame.quit()
        #     input("Are you ready for the next run? ")

        # --------------------------------- DEBUGGING ---------------------------------

        # Adding the best 12 agents to the next generation for crossover and mutation.
        for i in range(12):
            newGeneration.append(currentFitness[i][1])
        
        population = crossover(newGeneration)
        population = mutation(population)

        generation += 1
        end = time.time()
        
        # How many generations it's going to train for.
        if generation == numGens:
            return currentFitness

        print("Time for generation: {}".format(end - start))

def displayGame(snake, lastMove, agent, fruit):

    # Creating the screen.
    global res
    global screen
    background_colour = pygame.Color("#8fcb9e")
    res = (900, 800)
    screen = pygame.display.set_mode(res)
    pygame.display.set_caption('Snake Game')
    screen.fill(background_colour)
    pygame.display.flip()
    pygame.init()

    smallfont = pygame.font.SysFont('Corbel',35)
    clock = pygame.time.Clock()

    # Drawing the grid.
    for x in range(blockSize, res[0] - 20, blockSize):
        for y in range(blockSize, res[1] - 80, blockSize):
            pygame.draw.rect(screen, pygame.Color(GRASS), pygame.Rect(x, y, blockSize, blockSize))
            pygame.draw.rect(screen, BLACK, pygame.Rect(x, y, blockSize, blockSize), 1)

    # Drawing the snake's head.
    pygame.draw.rect(screen, BLACK, snake.draw())

    # Generating and drawing the fruit.
    fruit.generateFruit()
    fruit.drawFruit()

    # Head of snake.
    snake.body.append(pygame.Rect(snake.x, snake.y, blockSize, blockSize))

    # Tail of snake.
    snake.body.append(pygame.Rect(snake.x, snake.y, blockSize, blockSize))
    
    score = 0

    while True:

        # This is so the user can close the window anytime when displaying.
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                exit()

        # --------------------------------- DEBUGGING ---------------------------------
        # if clock.tick(6):

        #     newEvent = pygame.event.wait()
        #     if newEvent.type == pygame.KEYDOWN:
                        
        #         MoveSnake(lastMove, snake, True)
        #         move = makeMove(agent, snake, lastMove, fruit, True)

        #         print("Snake x: {}\tSnake y: {}\tFruit x: {}\tFruit y: {}".format(snake.x, snake.y, fruit.x, fruit.y))
        #         print()
        #         print("Wall: {}".format(move[1][0]))
        #         print("Current Direction: {}".format(move[1][1]))
        #         print("Relative X and Y: {}".format(move[1][2]))

        #         if move[0][0] == max(move[0]):
        #             if lastMove != "right":
        #                 lastMove = "left"
        #         if move[0][1] == max(move[0]):
        #             if lastMove != "left":
        #                 lastMove = "right"
        #         if move[0][2] == max(move[0]):
        #             if lastMove != "down":
        #                 lastMove = "up"
        #         if move[0][3] == max(move[0]):
        #             if lastMove != "up":
        #                 lastMove = "down"

        #         pygame.display.update()

        # --------------------------------- DEBUGGING ---------------------------------


        # Score variable.
        scoreTitle = smallfont.render('score:' , True , BLACK)
        scoreValue = smallfont.render(str(score), True, BLACK)
        screen.blit(scoreTitle, (center[0] - 75, center[1] + 330))
        screen.blit(scoreValue, (center[0] + 15, center[1] + 330))

        # This moves the snake at a certain time interval.
        if clock.tick(6):

            # Uses the trained agent to play the game with it's neural network. (Set to true for debugging).
            move = makeMove(agent, snake, lastMove, fruit, False)

            if move[0] == max(move):
                if lastMove != "right":
                    lastMove = "left"
            if move[1] == max(move):
                if lastMove != "left":
                    lastMove = "right"
            if move[2] == max(move):
                if lastMove != "down":
                    lastMove = "up"
            if move[3] == max(move):
                if lastMove != "up":
                    lastMove = "down"

            MoveSnake(lastMove, snake, True)
            pygame.display.update()

        # When the agent gets a fruit it will replace it with another randomly generated fruit.
        # Then it will add one to the score and display it.
        if snake.x == fruit.x and snake.y == fruit.y:

            # Adds a part to the body.
            Grow(snake, lastMove)

            # Generate a new fruit.
            fruit.generateFruit()
            fruit.drawFruit()

            # Increase score by one.
            score += 1

            # Erase the old score and put in the new score.
            scoreValue = smallfont.render(str(score), True, BLACK)
            screen.fill(background_colour, (center[0] + 15, center[1] + 330, 100, 100))
            screen.blit(scoreValue, (center[0] + 15, center[1] + 330))
        
        pygame.display.update()

        # If the snake hit's itself.
        if len(snake.body) > 2:
            for parts in snake.body[1:]:
                if snake.x == parts.x and snake.y == parts.y:
                    return

        # If the snake hits a wall on the horizontal axis.
        if snake.x < 20 or snake.x > 860:
            return

        # If the snake hits a wall on the vertical axis.
        elif snake.y < 20 or snake.y > 700:
            return

# Initilizing variables.
res = (900, 800)
width = res[0]
height = res[1]
center = (int(width / 2), int(height / 2))
lastMove = "left"
blockSize = 20

BLACK = [0, 0, 0]
RED = [255, 0, 0]
GRASS = ("#7EC984")
color = (255,255,255) 
color_light = (170,170,170) 
color_dark = (100,100,100)

amountOfChromosomes = 16321 

# Creating a population of agents.
population = []
for i in range(50):
    population.append(Agent(0, 0, 0, []))
    population[i].createPopulation()

# Run the training.
results = trainGen(population, 100)

# This is for when the training is completed it will wait for user input to display results.
x = input("are you ready to see the results?")

# Display the 20 best agents.
for i in range(20):
    print("Fitness: {}".format(results[i][0]))
    displayGame(Snake(random.randrange(20, 860, 20), random.randrange(20, 700, 20), []), "left", results[i][1], Fruit(0, 0))